apex/transformer/tensor_parallel/layers.py

```diff
@@ -313,45 +320,76 @@ def forward(
+                world_size = get_tensor_model_parallel_world_size()
+                shape = list(input.shape)
+                shape[0] *= world_size
+
+                all_gather_buffer = torch.empty(
+                    shape,
+                    dtype=input.dtype,
+                    device=torch.cuda.current_device(),
+                    requires_grad=False,
+                )
+                handle = torch.distributed._all_gather_base(
+                    all_gather_buffer,
+                    input,
+                    group=get_tensor_model_parallel_group(),
+                    async_op=True,
+                )
+                total_input = all_gather_buffer
+            else:
+                total_input = input
 
-            all_gather_buffer = torch.empty(
-                shape,
-                dtype=input.dtype,
-                device=torch.cuda.current_device(),
-                requires_grad=False,
-            )
-            handle = torch.distributed._all_gather_base(
-                all_gather_buffer,
-                input,
-                group=get_tensor_model_parallel_group(),
-                async_op=True,
-            )
-            total_input = all_gather_buffer
-        else:
-            total_input = input
         grad_input = grad_output.matmul(weight)
 
-        if ctx.sequence_parallel_enabled:
+        if handle is not None:
             handle.wait()
 
-        # Convert the tensor shapes to 2D for execution compatibility
-        grad_output = grad_output.view(
-            grad_output.shape[0] * grad_output.shape[1], grad_output.shape[2]
-        )
-        total_input = total_input.view(total_input.shape[0] * total_input.shape[1], total_input.shape[2])
         if ctx.async_grad_allreduce:
             # Asynchronous all-reduce
             handle = torch.distributed.all_reduce(
                 grad_input, group=get_tensor_model_parallel_group(), async_op=True
             )
 
+        #if no weight gradient, immediately return
+        if not ctx.compute_weight_gradient:
+            if ctx.sequence_parallel_enabled:
+                assert not ctx.async_grad_allreduce
+                world_size = get_tensor_model_parallel_world_size()
+                shape = list(grad_input.shape)
+                shape[0] //= world_size
+
+                sub_grad_input = torch.empty(torch.Size(shape), dtype=grad_input.dtype, device=torch.cuda.current_device(), requires_grad=False)
+                handle = torch.distributed._reduce_scatter_base(
+                    sub_grad_input,
+                    grad_input,
+                    group=get_tensor_model_parallel_group(),
+                    async_op=True
+                )
+                handle.wait()
+                return sub_grad_input, None, None, None, None, None, None
+            if ctx.async_grad_allreduce:
+                handle.wait()
+            return grad_input, None, None, None, None, None, None
```

Comment:
QQ: is this `grad_input` dependent on L.355 of `grad_input = grad_output.matmul(weight)` and L.362 when async_grad_allreduce? (I'm trying to understand why the view op on grad_output and total_input are moved

Reply:
The view on grad_output and total input are only used for calculating the weight gradient so if `weight.requires_grad=False` then they are not needed so I moved them after the return on L.385 so that views would not be executed. Maybe it would be better to keep the views as they are but wrap them in an `if ctx.compute_weight_grad`

Reply:
understood. when I first saw "immediately return", I wondered if we could put the no weight grads logic before with weight grads.